mlflow/spark.py

```diff
@@ -235,21 +240,29 @@ def log_model(
     try:
         spark_model.save(posixpath.join(model_dir, _SPARK_MODEL_PATH_SUB))
     except Py4JError:
-        return Model.log(
-            artifact_path=artifact_path,
-            flavor=mlflow.spark,
-            spark_model=spark_model,
-            conda_env=conda_env,
-            code_paths=code_paths,
-            dfs_tmpdir=dfs_tmpdir,
-            sample_input=sample_input,
-            registered_model_name=registered_model_name,
-            signature=signature,
-            input_example=input_example,
-            await_registration_for=await_registration_for,
-            pip_requirements=pip_requirements,
-            extra_pip_requirements=extra_pip_requirements,
-        )
+        artifact_repo = get_artifact_repository(artifact_uri=run_root_artifact_uri)
+        if (
+            isinstance(artifact_repo, DatabricksArtifactRepository)
+            and databricks_utils.is_mlflowdbfs_available()
+        ):
+            # TODO: test we're saving model metadata when mlflowdbfs is available.
+            _log_model_via_mlflowdbfs(artifact_repo, artifact_path, spark_model)
```

Comment:
If this procedure fails, we should raise a clear and actionable failure message. Rather than nesting try / catch blocks, I propose that we move this code out of the exception handler, since, `isinstance(artifact_repo, DatabricksArtifactRepository) == True` implies that `run_root_artifact_uri` is of the form `dbfs:/databricks/mlflow-tracking/...`, which can't be written to via `spark_model.save(model_dir)` anyway. The current try / except should be moved to an `else` block.